### PR TITLE
Recommendations about loopback prototype methods for custom end-points

### PR DESCRIPTION
**Pull request by JonForest**
I think we should favour prototype over static methods for Model endpoints where we are passing in an :id argument.

This PR is for discussion around that topic.

### CODE REVIEW

**Comment by ewen** [comment]
LGTM 👍 

I also think it's worth having a convention of assigning `this` in the first line of the function for prototype methods.

```
Worksite.prototype.startWork = function (data, cb) {
  const instance = this
   ...
}
```

**Comment by JonForest** [comment]
@ewen You think `instance` is better than the model name?
e.g.
```
const worksite = this
```
?

**Comment by ewen** [comment]
@JonForest I'm OK with either, I've been using `instance` I think cause it's always consistent.

**Comment by pnw** [comment]
I'd say only use an alias if you use a meaningful name, like `worksite` or `permit` or `notice` or something. Otherwise, `instance` and `this` are both equally as general/ambiguous and I'd rather just use `this`.

File: loopback-style-guide.md [hunk 1]
```diff
@@ -1,3 +1,69 @@
...
+ })
+
+ Worksite.prototype.startWork = function (data, cb) {
+   // 'this' -> is the worksite
+   ...
+```
+
+* In this example, `this` is the worksite, and a failure to find a worksite matching the supplied Id returns a 404 before it hits this method.
+* This mechanism reduces the code required in your Model methods.
+* **Note** It is important to use traditional function declarations over ES6 arrow functions when declaring your prototype, otherwise the `this` keyword will not have the correct context.
```
Comment (markstuart) [inline loopback-style-guide.md]: I have been burnt by this 🔥

File: loopback-style-guide.md [hunk 2]
```diff
@@ -1,3 +1,69 @@
...
+* In this example, we are required to look up the target model as one of the first operations of the method, and then check this was a valid model. 
+* All this can be handled for us if we use prototype methods, in the following structure:
+
+```javascript
+ Worksite.remoteMethod('startWork', {
+   accepts: [
+     {arg: 'id', type: 'number', required: true},
+     {arg: 'data', type: 'object', http: {source: 'body'}}
+   ],
+   isStatic: false,
```
Comment (markstuart) [inline loopback-style-guide.md]: Important to note that this line is the only difference between defining a static or instance remote method.

**Comment by markstuart** [comment]
Another thing to note is that the `slc` command line tools provide a pretty good custom remote method generator. It asks as part of the the interactive process if you want the method to be static or not, and generates the remote definition into the model JSON file. You then need to create the method handler in the model js file.

File: loopback-style-guide.md [hunk 3]
```diff
@@ -1,3 +1,69 @@
...
+   returns: {type: 'Worksite', root: true},
+   http: {path: '/start-work', verb: 'post', status: 200}
```
Comment (markstuart) [inline loopback-style-guide.md]: Oh, and of course the `/:id` portion of the path is removed.

**Comment by ewen** [comment]
@pnw I think it's more than just a naming thing, assigning it first removes any thinking about the `this` context later in the function. 

Agree on the naming - may as well make the convention a specific name rather than a generic one.

File: loopback-style-guide.md (hunk 2 above)
Comment (pnw) [inline loopback-style-guide.md]: 1. `isStatic: false`
2. the `/:id` portion of the path is removed like you said
3. The method must be a prototype method `Worksite.prototype.startWork` not `Worksite.startWork`

**Comment by anotheredward** [comment]
Sweet job, I've added the requested parts to the document, let's get this merged in!

File: loopback-style-guide.md [hunk 4]
```diff
@@ -1,3 +1,71 @@
...
+ })
+
+ Worksite.prototype.startWork = function (data, cb) {
+   const worksite = this
```
Comment (pnw) [inline loopback-style-guide.md]: Still not convinced by this convention. Is there actually any benefit to setting an alias like this besides mindless conformity? It's not consistent with how we use `this` in any other scenario in javascript.

@anotheredward

Comment (anotheredward) [inline loopback-style-guide.md]: Cool, would you like to revert the change and then we can merge this @pnw ? Just trying to get things cleaned up :)

Comment (ewen) [inline loopback-style-guide.md]: @pnw lets leave it out.

**Comment by anotheredward** [comment]
Cool, this has been discussed and agreed upon. Merging!
